Don't crash WebContent process when passing invalid URL for cache

chirags27 opened the pull request:
#### 99afca995277e6c9ac1e2e68ccfc133b4d99fc3d
<pre>
Don&apos;t crash WebContent process when passing invalid URL for cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=250727">https://bugs.webkit.org/show_bug.cgi?id=250727</a>
rdar://104294986

Reviewed by Chris Dumez and Sihui Liu.

Before this change, if FetchRequest::create() returned an exception, we
never used to check for that and call releaseReturnValue() on that which
would cause a crash. This change fixes that, so that we can correctly
return the &quot;URL is not valid or contains user credentials.&quot; exception.
This change also fixes a bug where correct exception wasn&apos;t returned.

* LayoutTests/http/tests/cache-storage/cache-match-invalid-url-expected.txt: Added.
* LayoutTests/http/tests/cache-storage/cache-match-invalid-url.html: Added.
* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::doMatch):
(WebCore::DOMCache::matchAll):
(WebCore::DOMCache::requestFromInfo):
* Source/WebCore/Modules/cache/DOMCache.h:

Canonical link: <a href="https://commits.webkit.org/259091@main">https://commits.webkit.org/259091@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88ad07ea4e2c754d070800991ac985f3ce4c4f53

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12927 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113036 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173347 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107760 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3822 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96058 "Built successfully") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112148 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109583 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10757 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93821 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38471 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92588 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25429 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80122 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6284 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26816 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6452 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3355 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12436 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46339 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6257 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8220 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (approved):
LGTM

szewai commented on LayoutTests/http/tests/cache/cache-match-invalid-url.html:
```diff
@@ -0,0 +1,19 @@
+<script>
```
This test should probably go into LayoutTests/http/tests/cache-storage, not cache/

szewai commented on LayoutTests/http/tests/cache/cache-match-invalid-url.html:
```diff
@@ -0,0 +1,19 @@
+
+function testFunc() {
+  caches.open('a').then((cache) => {
+    cache.match(`/\\X\u0600`);
```
We could try and catch here, and verify the exception is expected (e.g. url is invalid).
Maybe something like:
try {
        cache.match(`/\\X\u0600`);
        assert_unreached("Expected exception to be thrown");
} catch (err) {
        assert_equals(err.code, DOMException.INVALID_STATE_ERR, "The exception message");
}
Will need to import some helper js files as other tests in cache-storage/.

cdumez replied:
This function returns a promise, therefore, it will reject the promise, not throw an exception, unless we make the function async and use `await`.

chirags27 replied:
Good point, I'll update the test so that it can validate the exception

chirags27 replied:
Based on https://developer.mozilla.org/en-US/docs/Web/API/DOMException#error_names, the exception object is of the type TypeError, which is what the test validates.

szewai commented on LayoutTests/imported/w3c/web-platform-tests/service-workers/cache-storage/cache-match.https.any.js:
```diff
@@ -56,8 +56,11 @@ prepopulated_cache_test(simple_entries, function(cache, entries) {
       .then(function(result) {
-          assert_equals(result, undefined,
-                        'Cache.match should not match HEAD Request.');
+          assert_unreached('Should not reach here, exception expected')
```
This change does not seem right, according to https://w3c.github.io/ServiceWorker/#cache-matchall, this should return undefined when method check fails, not throwing an exception.

Note that the tests in LayoutTests/imported/w3c/web-platform-tests/ directory are imported from a public test set (https://web-platform-tests.org), which is usually compliant to standard, so you might want to double check the spec when your change breaks it.

chirags27 replied:
Makes sense, I'll fix the code to deal with this, thanks for pointing that out

szewai reviewed (changes requested):
I think we should find a way to not break existing imported tests.

szewai commented on LayoutTests/http/tests/cache-storage/cache-match-invalid-url.html:
```diff
@@ -0,0 +1,28 @@
+</script>
+<body onload="testFunc()">
+<p>
+PASS if no crash.
```
It might be better to use 'PASS Throwing exception on invalid url' to match the the test name.

szewai commented on Source/WebCore/Modules/cache/DOMCache.h:
```diff
@@ -66,7 +66,7 @@ class DOMCache final : public RefCounted<DOMCache>, public ActiveDOMObject {
     DOMCache(ScriptExecutionContext&, String&& name, DOMCacheIdentifier, Ref<CacheStorageConnection>&&);
 
-    ExceptionOr<Ref<FetchRequest>> requestFromInfo(RequestInfo&&, bool ignoreMethod);
+    ExceptionOr<Ref<FetchRequest>> requestFromInfo(RequestInfo&&, bool ignoreMethod, bool* requestValidationFailure = nullptr);
```
s/requestValidationFailure/requestValidationFailed

szewai reviewed (approved):
r=me

Commit:
Don't crash WebContent process when passing invalid URL for cache
https://bugs.webkit.org/show_bug.cgi?id=250727
rdar://104294986

Reviewed by Chris Dumez and Sihui Liu.

Before this change, if FetchRequest::create() returned an exception, we
never used to check for that and call releaseReturnValue() on that which
would cause a crash. This change fixes that, so that we can correctly
return the "URL is not valid or contains user credentials." exception.
This change also fixes a bug where correct exception wasn't returned.

* LayoutTests/http/tests/cache-storage/cache-match-invalid-url-expected.txt: Added.
* LayoutTests/http/tests/cache-storage/cache-match-invalid-url.html: Added.
* Source/WebCore/Modules/cache/DOMCache.cpp:
(WebCore::DOMCache::doMatch):
(WebCore::DOMCache::matchAll):
(WebCore::DOMCache::requestFromInfo):
* Source/WebCore/Modules/cache/DOMCache.h:

Canonical link: https://commits.webkit.org/259091@main
